upstairs/src/impacted_blocks.rs

```diff
@@ -2,96 +2,237 @@
 
-        false
+                // If first == last, then the iterator is now done. Otherwise,
+                // increment first.
+                if first == last {
+                    self.active_range = None
+                } else if last.block == 0 {
+                    last.block = self.extent_size - 1;
+                    last.extent_id -= 1;
+                } else {
+                    last.block -= 1;
+                }
+
+                Some(result)
+            }
+        }
     }
+}
+
+impl FusedIterator for ImpactedBlockIter {}
 
-    /// Return a list of impacted extents
-    #[cfg(test)]
-    pub fn extents(&self) -> Vec<&u64> {
-        self.blocks.keys().collect()
+impl ImpactedBlocks {
+    /// Create a new ImpactedBlocks range from first and last impacted blocks
+    /// (inclusive) Returns Empty if last_impacted < first_impacted.
+    pub fn new(
+        first_impacted: ImpactedAddr,
+        last_impacted: ImpactedAddr,
+    ) -> Self {
+        if first_impacted <= last_impacted {
+            ImpactedBlocks::InclusiveRange(first_impacted, last_impacted)
+        } else {
+            ImpactedBlocks::Empty
+        }
     }
 
-    /// Returns (extent id, block) tuples
-    pub fn tuples(&self) -> Vec<(u64, Block)> {
-        let mut result = Vec::with_capacity(self.len());
+    /// Create a new ImpactedBlocks range starting at a given offset, and
+    /// stretching n_blocks further into the extent. Panics an error if
+    /// `extent_size` is 0. Returns ImpactedBlocks::Empty if n_blocks is 0.
+    pub fn from_offset(
+        extent_size: u64,
+        first_impacted: ImpactedAddr,
+        n_blocks: u64,
+    ) -> Self {
+        debug_assert!(extent_size > 0);
 
-        let sorted_keys = self.blocks.keys().sorted();
-        for eid in sorted_keys {
-            let block_offsets =
-                self.blocks[eid].iter_ones().collect::<Vec<usize>>();
-            for i in block_offsets.iter().sorted() {
-                result
-                    .push((*eid, Block::new_with_ddef(*i as u64, &self.ddef)));
-            }
+        if n_blocks == 0 {
+            return ImpactedBlocks::Empty;
         }
 
-        result
+        // So because we're inclusive, if we have 1 block then the first impacted should be the same as the last impacted. If we have 2 blocks, it's +1. etc.
+        let ending_block = n_blocks + first_impacted.block - 1;
+
+        let last_impacted = ImpactedAddr {
+            extent_id: first_impacted.extent_id + ending_block / extent_size,
+            block: ending_block % extent_size,
+        };
+
+        ImpactedBlocks::InclusiveRange(first_impacted, last_impacted)
     }
 
-    /// Returns the number of impacted blocks
-    pub fn len(&self) -> usize {
-        let mut len = 0;
+    pub fn intersection(&self, other: &ImpactedBlocks) -> ImpactedBlocks {
+        let ImpactedBlocks::InclusiveRange(fst_self, lst_self) = self else {
```

Comment:
TIL about the `let else`. Nice.

Reply:
yeah it got added into rust stable recently and it is suuuper helpful for a lot of common patterns